### PR TITLE
Dotp

### PR DESCRIPTION
**Pull request by mbertuletti**
Summarize the changes concisely

### Added
Added parallel dotp functions with different reduction approaches

## Checklist

- [x] Automated tests pass
- [x] Changelog updated
- [x] Code style guideline is observed

### CODE REVIEW

File: hardware/deps/snitch/src/snitch.sv [hunk 1]
```diff
@@ -290,6 +282,51 @@ module snitch
...
+        stall_lsu_q <= stall_lsu_q + 1;
+      end
+
+    end
+
+  end
+
+  `endif
+
+
```
Comment (SamuelRiedel) [inline hardware/deps/snitch/src/snitch.sv:329]: The functional point looks good to me. Only the naming of `stall_instret_q` is a bit confusing. `instret_q` stands for instruction retired, but the `stall_instret_q` measures instruction cache stalls, right? Maybe you can rename this and add a comment with a quick description of what the counters do.

I have some tips about our coding style as well. Can you please remove all the empty lines between the code blocks and remove the duplicate _performance counter_ comment at the top?

Also, always ensure each signal has a value assigned for each case. For example, specify a default like `stall_raw_q <= stall_raw_q;` and then apply your conditional update.

Finally, to simplify everything, you should consider just writing each register with the ``` `FFLAR``` macros. They already handle all those nitpicks I just mentioned. You could rewrite the whole `always_ff` block as follows:
```
`FFLAR(stall_instret_q, stall_instret_q + 1, stall && !inst_ready_i && inst_valid_o, '0, clk_i, rst_i);
`FFLAR(stall_raw_q, stall_raw_q + 1, !operands_ready || !dst_ready, '0, clk_i, rst_i);
`FFLAR(stall_lsu_q, stall_lsu_q + 1, lsu_stall, '0, clk_i, rst_i);
```

File: hardware/deps/snitch/src/snitch.sv [hunk 2]
```diff
@@ -1369,6 +1406,7 @@ module snitch
     // Right now we skip this due to simplicity.
     if (csr_en) begin
       unique case (inst_data_i[31:20])
+
```
Comment (SamuelRiedel) [inline hardware/deps/snitch/src/snitch.sv:1409]: Again you have some unnecessary whitespace changes.
```suggestion
```

File: hardware/deps/snitch/src/snitch.sv [hunk 3]
```diff
@@ -1389,6 +1427,16 @@ module snitch
...
+        riscv_instr::CSR_MHPMCOUNTER3: begin
+          csr_rvalue = stall_instret_q[31:0];
+        end
+        riscv_instr::CSR_MHPMCOUNTER4: begin
+          csr_rvalue = stall_lsu_q[31:0];
+        end
+        riscv_instr::CSR_MHPMCOUNTER5: begin
+          csr_rvalue = stall_raw_q[31:0];
+        end
+
```
Comment (SamuelRiedel) [inline hardware/deps/snitch/src/snitch.sv:1439]: The same with whitespace changes here. Also, you have 64-bit counters, but we can read only 32 bits. Shall we simply make them 32-bit counters?

File: software/apps/dotp/main.c [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+#include "dotp_parallel_local.h"
+#endif
+
+#if defined(PARALLEL_RED0) || defined(PARALLEL_UNROLLED_RED0)
+#include "dotp_parallel_red0.h"
+#endif
+
+#if defined(PARALLEL_REDTREE) || defined(PARALLEL_UNROLLED_REDTREE)
+#include "dotp_parallel_redtree.h"
+#endif
```
Comment (SamuelRiedel) [inline software/apps/dotp/main.c:36]: If I am not mistaken, all the header files define different functions, so you could technically merge them into a single header or always include all of them.

Or you could have all of them have the same function names and signatures and then simplify the `main.c` to include the one you want, and the remainder of the main would be the same for all of them. At the moment, you include different header files, and then you call different functions depending on the includes, so you have the duplication twice.

For me, it's fine as is, but I just wanted to mention it in case it is possible to do that.

File: software/apps/dotp/main.c [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+#ifdef SINGLE
+  time_init = mempool_get_timer();
+  dotp_single(vector_a, vector_b, &sum, LEN);
+  time_end = mempool_get_timer();
+#elif defined(SINGLE_UNROLLED)
+  time_init = mempool_get_timer();
+  dotp_single_unrolled4(vector_a, vector_b, &sum, LEN);
+  time_end = mempool_get_timer();
+#endif
```
Comment (SamuelRiedel) [inline software/apps/dotp/main.c:92]: Here, and in the following blocks, you have a lot of unnecessary duplication of the timer and benchmarking statements. You could potentially have those statements only once, and then one big switch case out of the defines with the function calls.

```suggestion
time_init = mempool_get_timer();
#ifdef SINGLE
  dotp_single(vector_a, vector_b, &sum, LEN);
#elif defined(SINGLE_UNROLLED)
  dotp_single_unrolled4(vector_a, vector_b, &sum, LEN);
#endif
time_end = mempool_get_timer();
```

**Review by SamuelRiedel — CHANGES_REQUESTED** [review]
Thanks, Marco. In general, it looks very good to me. I just added some long explanations on our coding guidelines for HW and a few tips to simplify the SW. Please fix the HW parts, but whether and how much you want to modify the SW is up to you.

File: hardware/deps/snitch/src/snitch.sv (hunk 1 above)
Comment (mbertuletti) [reply hardware/deps/snitch/src/snitch.sv]: Done.

File: hardware/deps/snitch/src/snitch.sv (hunk 2 above)
Comment (mbertuletti) [reply hardware/deps/snitch/src/snitch.sv]: Done.

File: hardware/deps/snitch/src/snitch.sv (hunk 3 above)
Comment (mbertuletti) [reply hardware/deps/snitch/src/snitch.sv]: Done.
